### PR TITLE
Add lintr hook

### PR DESCRIPTION
**Pull request by jonmcalder**
@lorenzwalthert could you check/test this please? For now I've just attempted to implement a basic check for lint free files to first see if I've got the process correct.

I think we'll ultimately want to support arguments to allow for customizing what linting options are applied.

### CODE REVIEW

**Comment by lorenzwalthert** [comment]
When you add `lintr`  as a dependency in DESCRIPTION, the CI should pass.

File: .pre-commit-hooks.yaml [hunk 1]
```diff
@@ -46,4 +46,10 @@
...
     language: script
-    files: '\.[rR]$'
\ No newline at end of file
+    files: '\.[rR]$'
+-   id: lintr
+    name: lintr
+    description: check if a .R file is lint free (using lintr)
+    entry: bin/lintr
+    language: script
+    files: '\.R$'
```
Comment (lorenzwalthert) [inline .pre-commit-hooks.yaml]: Please use `'\.[rR]$'`.

File: tests/testthat/test-all.R [hunk 2]
```diff
@@ -119,3 +119,15 @@ test_temp(suffix = "-success.R", error_msg = NULL)
 
 # fail (call to library that is not in description)
 test_temp(suffix = "-fail.R", error_msg = "Dependency check failed")
+
+### . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ..
+### lintr                                                                   ####
+
+# success
+run_test("lintr",
```
Comment (lorenzwalthert) [inline tests/testthat/test-all.R]: Please reformat with styler.

**Comment by lorenzwalthert** [comment]
What kind of options do you mean? Do you mean how to make use of `.lintr`  file or pass arguments to `lintr::lint()`?

**Comment by jonmcalder** [comment]
> What kind of options do you mean? Do you mean how to make use of `.lintr` file or pass arguments to `lintr::lint()`?

Initially I was just thinking to pass arguments to `lintr::lint()`, but I guess whatever approach would be easier and/or better? What do you think?

**Comment by lorenzwalthert** [comment]
I don't know lintr very well to be honest. How would you specify the linters in a `lintr::lint()`  call? LIke this: 

```r
lint("R/data.R", list(assignment_linter, commented_code_linter))
```
So these must be passed throught the `.pre-commit-config.yaml` to the `lintr::lint()`  call. I think theis can become a bit messy if there are many.
pre-commit wants you to commit the `.pre-commit-config.yaml`  whenever you change it right away, which is a good way to ensure that the tests passed by the hook correspond to the config of the hook. With lintr, you could for example remove all rules from the .lintr file without committing the .lintr file, and then all your files would pass the hook. You essentially had to *always* commit your .lintr changes when you commit anything. This could be ensured by checking from the hook if there is a .lintr file in the staging area and let the hook fail if that is the case. Not sure if we'd have to use git2r for this or if we can use git via bash< from R.

File: bin/lintr [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env Rscript
+files <- commandArgs(trailing = TRUE)
+
+if (git2r::in_repository()) {
+  git_status <- git2r::status(staged = FALSE)
+  if (any(grepl("\\.lintr", git_status))) {
```
Comment (lorenzwalthert) [inline bin/lintr]: I think we could limit this to the `.lintr` file in the current working directory and flatten a potentially nested list (e.g. git_status$unstaged$modified`) first.
```suggestion
  if (any(grepl(".lintr", unlist(git_status), fixed = TRUE))) {
```

File: bin/lintr [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    stop("Unstaged changes to .lintr file", call. = FALSE)
```
Comment (lorenzwalthert) [inline bin/lintr]: Should we make this a bit more verbose?
```suggestion
    stop("Unstaged changes to .lintr file. Stage the .lintr file or discard the changes to it. ", call. = FALSE)
```

**Comment by lorenzwalthert** [comment]
Thanks Jon for going the extra mile with `git2r`. What do you think about my two minor suggestions?

**Comment by lorenzwalthert** [comment]
Let's merge this, thanks Jon.

**Pull request merged**